homeassistant/components/lastfm/sensor.py

```diff
@@ -35,95 +30,49 @@
+    def __init__(self, coordinator: LastFmUpdateCoordinator, user: str) -> None:
         """Initialize the sensor."""
-        self._unique_id = hashlib.sha256(user.encode("utf-8")).hexdigest()
-        self._user = lastfm_api.get_user(user)
-        self._name = user
-        self._lastfm = lastfm_api
-        self._state = "Not Scrobbling"
-        self._playcount = None
-        self._lastplayed = None
-        self._topplayed = None
-        self._cover = None
-
-    @property
-    def unique_id(self):
-        """Return the unique ID of the sensor."""
-        return self._unique_id
-
-    @property
-    def name(self):
-        """Return the name of the sensor."""
-        return self._name
-
-    @property
-    def native_value(self):
-        """Return the state of the sensor."""
-        return self._state
-
-    def update(self) -> None:
-        """Update device state."""
-        self._cover = self._user.get_image()
-        self._playcount = self._user.get_playcount()
-
-        if recent_tracks := self._user.get_recent_tracks(limit=2):
-            last = recent_tracks[0]
-            self._lastplayed = f"{last.track.artist} - {last.track.title}"
-
-        if top_tracks := self._user.get_top_tracks(limit=1):
-            top = str(top_tracks[0])
-            if (toptitle := re.search("', '(.+?)',", top)) and (
-                topartist := re.search("'(.+?)',", top)
-            ):
-                self._topplayed = f"{topartist.group(1)} - {toptitle.group(1)}"
-
-        if (now_playing := self._user.get_now_playing()) is None:
-            self._state = STATE_NOT_SCROBBLING
-            return
-
-        self._state = f"{now_playing.artist} - {now_playing.title}"
-
-    @property
-    def extra_state_attributes(self):
-        """Return the state attributes."""
-        return {
-            ATTR_LAST_PLAYED: self._lastplayed,
-            ATTR_PLAY_COUNT: self._playcount,
-            ATTR_TOP_PLAYED: self._topplayed,
-        }
-
-    @property
-    def entity_picture(self):
-        """Avatar of the user."""
-        return self._cover
+        super().__init__(coordinator)
+        self._attr_unique_id = hashlib.sha256(user.encode("utf-8")).hexdigest()
+        self._attr_name = f"lastfm_{user}"
+        self._user = user
+
+    @callback
+    def _handle_coordinator_update(self) -> None:
+        if user_data := self.coordinator.data.get(self._user):
+            self._attr_entity_picture = user_data.image
+            self._attr_native_value = user_data.now_playing
+            self._attr_extra_state_attributes = {
+                ATTR_LAST_PLAYED: user_data.last_played,
+                ATTR_PLAY_COUNT: user_data.play_count,
+                ATTR_TOP_PLAYED: user_data.top_played,
+            }
+        super()._handle_coordinator_update()
+
+    async def async_added_to_hass(self) -> None:
+        """Subscribe to updates."""
+        self._handle_coordinator_update()
+        await super().async_added_to_hass()
```

Comment:
```suggestion
    def update(self) -> None:
        """Update device state."""
        self._attr_entity_picture = lastfm_user.get_image(SIZE_SMALL)
        if lastfm_user.get_now_playing() is not None:
            self._attr_native_value = format_track(self._user.get_now_playing())
        else:
            self._attr_native_value = STATE_NOT_SCROBBLING
        top_played = None
        if top_tracks := self._user.get_top_tracks(limit=1):
            top_played = format_track(top_tracks[0].item)
        last_played = format_track(
            self._user.get_recent_tracks(limit=1)[0].track
        )
        play_count = self._user.get_playcount()
        self._attr_extra_state_attributes = {
            ATTR_LAST_PLAYED: last_played,
            ATTR_PLAY_COUNT: play_count,
            ATTR_TOP_PLAYED: top_played,
        }
```

Reply:
I am wondering, why the sync update and not the async update? What's the reason behind that?

Reply:
Sync means that HA automatically runs it inside a `add_executor_job` for you.
If you go async then you need to handle that.